### ISSUE TITLE
Support Native Map Type

### ISSUE THREAD

**Issue by svozza** [sanctuary-js/sanctuary#233]
Someone recently commented on an old Ramda [issue](https://github.com/ramda/ramda/issues/1055) about supporting the native `Map` type and it got me thinking whether we should do so in Sanctuary. I'm not quite sure how that would look given that the `StrMap` type we have is more specific that the general JS type but I'm interested to hear everyone's thoughts.

**Comment by davidchambers** [comment]
I like the idea of a `Map k v` type, but I'm not sure the built-in Map type makes sense in a world of value-based equality. Consider this example:

```javascript
> new Map([[[1, 2, 3], 'foo'], [[1, 2, 3], 'bar']])
Map { [ 1, 2, 3 ] => 'foo', [ 1, 2, 3 ] => 'bar' }
```

`Map k v` and `StrMap v` can happily coexist. In fact, `StrMap v` and `Map String v` are *conceptually* equivalent.

**Comment by svozza** [comment]
Yes. I see what you mean. I guess one way to go would be to change our `StrMap` functions to also accept `Map String v` objects. Another would be to only support Maps that use primitive values as keys.

**Comment by davidchambers** [comment]
I'm happy to leave the role of working with wonky JavaScript constructs to Ramda. It seems inevitable to me that we will eventually provide `sanctuary-list`, `sanctuary-map`, and `sanctuary-set` packages to provide "immutable" data structures with value-based equality semantics.

Friday will be my last day at Plaid. I then intend to be unemployed for three months while my girlfriend and I decide where to live in Europe, so I should have time to create these packages. :)

🇩🇪 🇸🇪 🇳🇱

**Comment by davidchambers** [comment]
I posted a question on Stack Overflow: [Custom equality semantics for Immutable.js data structures][1]. After writing the post I'm almost convinced we could get our desired behaviour by wrapping the native Map type. It could look something like this:

    Map k v = { negativeZero :: Maybe v
              , value :: NativeMap k v
              , fantasy-land/equals :: Map k v ~> Map k v -> Boolean
              , fantasy-land/map :: Map k v ~> (k -> a) -> Map a v
              , fantasy-land/bimap :: Map k v ~> (k -> a, v -> b) -> Map a b
              , ...
              }

@scott-christopher wrapped the native Set type in ramda/ramda#1639 (for internal use only). Is there a reason we cannot use a similar approach for externally visible maps and sets in Sanctuary?


[1]: http://stackoverflow.com/q/39557342/312785

**Comment by scott-christopher** [comment]
> Is there a reason we cannot use a similar approach for externally visible maps and sets in Sanctuary?

The only issue I see is that we would probably want to treat it as an immutable structure, which will end up with a bunch of instance cloning to work around the underlying mutable structure.

I've a got a bit of time free over the next week or two and was already looking at implementing some persistent Map and Set types to get a better understanding of a paper I saw recently, so providing that goes to plan they could potentially live under the Sanctuary banner if there's interest.

**Comment by davidchambers** [comment]
> if there's interest

There's most definitely interest! :)

We get into all sorts of trouble treating arrays as sets in Ramda. Once we have access to a Set type we'll be able to define clean versions of functions such as `keys`, `values`, and `intersection`.

**Comment by davidchambers** [comment]
@scott-christopher, I created [sanctuary-js/sanctuary-map][1] and [sanctuary-js/sanctuary-set][2]. You're welcome to work on a feature branch of one of these repositories.


[1]: https://github.com/sanctuary-js/sanctuary-map
[2]: https://github.com/sanctuary-js/sanctuary-set

**Comment by scott-christopher** [comment]
I've pushed some code to https://github.com/sanctuary-js/sanctuary-set/tree/pre-0.0.1 which contains an initial pass of implementing two variants of immutable sets, one based on balanced AVL trees and the other is specialised for Int32 values using Patricia tries. It is severely lacking in documentation and the code could do with a good clean-up and sanity check, though I figured it would be good to get some eyes over it sooner rather than later.

I've also got some Map versions of the same implementations which I'll tidy up a bit over the coming days and push up to the other repository.
